replica/tablets.cc

```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+    throw_with_backtrace<std::bad_function_call>();
+}
+
+class tablet_incremental_selector : public sstables::incremental_selector_impl {
+    const tablet_sstable_set& _tset;
+
+    // _cur_set and _cur_selector contain a snapshot
+    // for the currently selected compaction_group.
+    lw_shared_ptr<sstables::sstable_set> _cur_set;
+    std::optional<sstables::sstable_set::incremental_selector> _cur_selector;
+    dht::token _lowest_next_token = dht::maximum_token();
+
+public:
+    tablet_incremental_selector(const tablet_sstable_set& tset)
+            : _tset(tset)
+    {}
+
+    virtual std::tuple<dht::partition_range, std::vector<sstables::shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        // Always return minimum singular range, such that incremental_selector::select() will always call this function,
+        // which in turn will find the next sstable set to select sstables from.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::min());
+
+        if (!_cur_selector) {
+            auto idx = _tset.group_of(pos.token());
+            _cur_set = _tset._sstable_sets[idx];
+            // Set the next token to point to the possible storage group.
+            // It will be considered later on when the _cur_set is exhausted
+            _lowest_next_token = find_lowest_next_token(_tset.first_token_of(idx+1));
+            if (!_cur_set) {
+                auto lowest_next_position = _lowest_next_token.is_maximum()
+                    ? dht::ring_position_ext::max()
+                    : dht::ring_position_ext::starting_at(_lowest_next_token);
+                tablet_logger.debug("tablet_incremental_selector {}.{}: select pos={}: returning 0 sstables, next_pos={}",
+                        _tset.schema()->ks_name(), _tset.schema()->cf_name(), pos, lowest_next_position);
+                return std::make_tuple(std::move(current_range), std::vector<sstables::shared_sstable>{}, lowest_next_position);
+            }
+
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+        }
+
+        auto res = _cur_selector->select(pos);
+        // Return all sstables selected on the requested position from the first matching sstable set.
+        // This assumes that the underlying sstable sets are disjoint in their token ranges so
+        // only one of them contain any given token.
+        auto sstables = std::move(res.sstables);
+        // Return the lowest next position, such that this function will be called again to select the
+        // lowest next position from the selector which previously returned it.
+        // Until the current selector is exhausted. In that case,
+        // jump to the next compaction_group sstable set.
+        dht::ring_position_ext lowest_next_position = res.next_position;
+        if (lowest_next_position.is_max()) {
+            _cur_set = {};
```

Comment:
lowest_next_position.is_max() means we only switch selectors when last token is exhausted, is_max() refers to maximum token, not maximum token of current group.

also we need to do the check before selecting() as there's no guarantee last token of a group exists, and select() is only called after with a token of next group. and select() must be able to detect it and advance the selector before selecting.

Reply:
> lowest_next_position.is_max() means we only switch selectors when last token is exhausted, is_max() refers to maximum token, not maximum token of current group.

`lowest_next_position` is the `next_position` returned from the current selector.
It means that this selector is exhausted and we need to switch to the next one.
We return it when there are no more sstables matching the selected position, not the this position is last/maximum.
Or maybe I don't follow what you mean above.

Reply:
I think "if (lowest_next_position.is_max()) {" will not work for switching selectors.

And you can't reset current selector after _cur_selector->select(pos), you have to do it before. 

You should check: does the current position span the current selector or should I move to the next one?

And only then call _cur_selector->select(pos).

You cannot predict which tokens will come next. next_position is only used as a hint for the reader to know when to invoke the selector again, after current position is >= next_position.

Reply:
see this: https://github.com/scylladb/scylladb/pull/13666/commits/dc1c0708852646bc17e708bdad8dd7b96b2293f4#diff-970063d9f766ba24975f1955091d6bd0b071932333c47936fe2b4ff62032c580R2951

Reply:
I understand now what you do :-) You're waiting for underlying selector to be exhausted, which is communicated with ring_position_view::max(). I think that could work too. I still think it would be good that we add a test for this, even though it's indirectly tested, edge cases may be hard to catch with tests that are not specialized. We don't have to do it now, but it would be good to track it in a github issue. We can easily even reuse existing tests that were written as bugs were found.